RTCConfiguration: iceCandidatePoolSize. Cite spec and expand coverage

rwaldron opened the pull request:
Signed-off-by: Rick Waldron <waldron.rick@gmail.com>

<!-- Reviewable:start -->

<!-- Reviewable:end -->

rwaldron committed:
RTCConfiguration: iceCandidatePoolSize. Cite spec and expand coverage

Signed-off-by: Rick Waldron <waldron.rick@gmail.com>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/230905437)

# Firefox (nightly) #
Testing web-platform-tests at revision a80580359f9848b5f72551f056230ed078bec12a
Using browser at version BuildID 20170509100208; SourceStamp b21b974d60d3075ae24f6fb1bae75d0f122f28fc
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5876/webrtc/RTCConfiguration-iceCandidatePoolSize.html">/webrtc/RTCConfiguration-iceCandidatePoolSize.html</a></summary>


|                                    Subtest                                    | Results |                                                                 Messages                                                                 |
|-------------------------------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                               | OK      |                                                                                                                                          |
| `new RTCPeerConnection({ iceCandidatePoolSize }) with valid value: 0`         | FAIL    | `assert_equals: expected (number) 0 but got (undefined) undefined`                                                                       |
| `new RTCPeerConnection({ iceCandidatePoolSize }) with valid value: 127`       | FAIL    | `assert_equals: expected (number) 127 but got (undefined) undefined`                                                                     |
| `new RTCPeerConnection({ iceCandidatePoolSize }) with valid value: 255`       | FAIL    | `assert_equals: expected (number) 255 but got (undefined) undefined`                                                                     |
| `new RTCPeerConnection({ iceCandidatePoolSize }) with valid value: null`      | FAIL    | `assert_equals: expected (number) 0 but got (undefined) undefined`                                                                       |
| `new RTCPeerConnection({ iceCandidatePoolSize }) with valid value: undefined` | FAIL    | `assert_equals: expected (number) 0 but got (undefined) undefined`                                                                       |
| `new RTCPeerConnection({ iceCandidatePoolSize }) with invalid value: -1`      | FAIL    | `assert_throws: function "() => {\n        new RTCPeerConnection({\n          iceCandidatePoolSize\n        });\n      }" did not throw` |
| `new RTCPeerConnection({ iceCandidatePoolSize }) with invalid value: 256`     | FAIL    | `assert_throws: function "() => {\n        new RTCPeerConnection({\n          iceCandidatePoolSize\n        });\n      }" did not throw` |
| `setConfiguration({ iceCandidatePoolSize }) to a valid value: 0`              | FAIL    | `pc.setConfiguration is not a function`                                                                                                  |
| `setConfiguration({ iceCandidatePoolSize }) to a valid value: 127`            | FAIL    | `pc.setConfiguration is not a function`                                                                                                  |
| `setConfiguration({ iceCandidatePoolSize }) to a valid value: 255`            | FAIL    | `pc.setConfiguration is not a function`                                                                                                  |
| `setConfiguration({ iceCandidatePoolSize }) to an invalid value: -1`          | PASS    |                                                                                                                                          |
| `setConfiguration({ iceCandidatePoolSize }) to an invalid value: 256`         | PASS    |                                                                                                                                          |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/230905446)

# Chrome (unstable) #
Testing web-platform-tests at revision a80580359f9848b5f72551f056230ed078bec12a

foolip commented:
@soareschen, can you review?

soareschen commented on webrtc/RTCConfiguration-iceCandidatePoolSize.html:
```diff
@@ -1,35 +1,87 @@
+  // [EnforceRange]
+  reconfigureWithInvalidPoolSize(-1);
+  reconfigureWithInvalidPoolSize(256);
+
```
Should also test with `NaN` and +/- `Infinity`, which are explicitly checked by `EnforceRange`. 

A few additional check on invalid types such as object and string might be helpful as well.

rwaldron replied:
We're not testing `[EnforceRange]` itself, we only need to ensure that implementations have correctly implemented this configuration property with the specified integer range semantics.

soareschen replied:
Ok. I got confused because the comment line above seem to imply that it is testing the behavior of EnforceRange.

soareschen commented on webrtc/RTCConfiguration-iceCandidatePoolSize.html:
```diff
@@ -1,35 +1,87 @@
+  /*
+  setConfiguration({ iceCandidatePoolSize })
+  */
+  reconfigureWithValidPoolSize(0, 0);
```
The test parameters can be deduplicated by having a common test function that take in the parameter and call `initializeWithInvalidPoolSize` and `reconfigureWithValidPoolSize` respectively.

rwaldron replied:
(copied from below)

I would actually prefer to avoid introducing additional testing mechanism logic whenever possible. TBH, I would prefer if the function abstractions didn't exist at all, and that the entirety of this file was structured similar to these examples: 

- https://github.com/w3c/web-platform-tests/commit/7f89c4349a940ad78d87a86e6a6bbd6ca5415921
- https://github.com/w3c/web-platform-tests/commit/79092e1bb755102e9e387dfa89ce676794226b1a
- https://github.com/w3c/web-platform-tests/commit/044ae0edd0da2320150b7b6edbdb1a49e9a026a0
- https://github.com/w3c/web-platform-tests/commit/d94b87c3b1ba302502155c0e364cee11c4eec393
- https://github.com/w3c/web-platform-tests/commit/348334c42edf5d77028c765412d3958f5d1ceffb

soareschen replied:
Got it then. I guess if there is not too many permutation of parameters that is a better way to go.

soareschen commented on webrtc/RTCConfiguration-iceCandidatePoolSize.html:
```diff
@@ -1,35 +1,87 @@
-testInvalidPoolSize(-1)
-testInvalidPoolSize(1000)
-testInvalidPoolSize("string")
+  function initializeWithValidPoolSize(iceCandidatePoolSize, expectedSize) {
```
Nit: I find function names like `testInitializeWithValidPoolSize` to be more descriptive than `initializeWithValidPoolSize` on what the function is doing.

soareschen commented on webrtc/RTCConfiguration-iceCandidatePoolSize.html:
```diff
@@ -1,35 +1,87 @@
+
+  // Defaulting to 0
+  initializeWithValidPoolSize(null, 0);
+  initializeWithValidPoolSize(undefined, 0);
```
Is there a reason null and undefined are not tested for reconfiguration case? The check is done on WebIDL level so shouldn't matter if it is reconfigured with the same value.

rwaldron replied:
These tests have already revealed that the implementation in Firefox Nightly is not throwing on construction, but is throwing on reconfiguration.

rwaldron replied:
I would actually prefer to avoid introducing additional testing mechanism logic whenever possible. TBH, I would prefer if the function abstractions didn't exist at all, and that the entirety of this file was structured similar to these examples: 

- https://github.com/w3c/web-platform-tests/commit/7f89c4349a940ad78d87a86e6a6bbd6ca5415921
- https://github.com/w3c/web-platform-tests/commit/79092e1bb755102e9e387dfa89ce676794226b1a
- https://github.com/w3c/web-platform-tests/commit/044ae0edd0da2320150b7b6edbdb1a49e9a026a0
- https://github.com/w3c/web-platform-tests/commit/d94b87c3b1ba302502155c0e364cee11c4eec393

soareschen replied:
I guess the original rationale is that AFAIK idlharness do not test for dictionary passed to functions and instead only check for attributes of existing objects. But it's fine if the convention is not to test too much into WebIDL.

foolip replied:
FWIW, I think it's fine to test things that are part of Web IDL, because often it can't be tested by idlharness.js because it can't assume anything about what getters and setters should do.

rwaldron committed:
Eliminate testing function abstraction

- easier to track down failing tests
- enforcing correct failures on setConfiguration (the TypeError alone is not enough to determine a correct failure)

Signed-off-by: Rick Waldron <waldron.rick@gmail.com>

soareschen commented on webrtc/RTCConfiguration-iceCandidatePoolSize.html:
```diff
@@ -1,35 +1,109 @@
+*/
+
+test(() => {
+  assert_equals(typeof pc.setConfiguration, "function", "RTCPeerConnection.prototype.setConfiguration is not implemented");
```
Nit: It is probably more clear if we merge this test with the previous test, since the first success case would have throw and abort the remaining assertions. But I get why having the tests done separately is clearer in most other cases, so either way is fine.

soareschen commented:
LGTM :+1:

rwaldron commented:
Rebased

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/234006818)

# Sauce (safari) #
Testing web-platform-tests at revision 561d24fe186f9676430d2509d34523af7e32bf0c
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5876/webrtc/RTCConfiguration-iceCandidatePoolSize.html">/webrtc/RTCConfiguration-iceCandidatePoolSize.html</a></summary>


|                                      Subtest                                       | Results |                                                                                                                      Messages                                                                                                                      |
|------------------------------------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                    | ERROR   |                                                                                                                                                                                                                                                    |
| `Initialize a new RTCPeerConnection with no iceCandidatePoolSize`                  | FAIL    | `Can't find variable: RTCPeerConnection`                                                                                                                                                                                                           |
| `Initialize a new RTCPeerConnection with iceCandidatePoolSize: 0`                  | FAIL    | `Can't find variable: RTCPeerConnection`                                                                                                                                                                                                           |
| `Initialize a new RTCPeerConnection with iceCandidatePoolSize: 255`                | FAIL    | `Can't find variable: RTCPeerConnection`                                                                                                                                                                                                           |
| `Initialize a new RTCPeerConnection with iceCandidatePoolSize: -1 (Out Of Range)`  | FAIL    | `assert_throws: function "() => {\n    new RTCPeerConnection({\n      iceCandidatePoolSize: -1\n    });\n  }" threw object "ReferenceError: Can't find variable: RTCPeerConnection" ("ReferenceError") expected object "TypeError" ("TypeError")`  |
| `Initialize a new RTCPeerConnection with iceCandidatePoolSize: 256 (Out Of Range)` | FAIL    | `assert_throws: function "() => {\n    new RTCPeerConnection({\n      iceCandidatePoolSize: 256\n    });\n  }" threw object "ReferenceError: Can't find variable: RTCPeerConnection" ("ReferenceError") expected object "TypeError" ("TypeError")` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/234006825)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 561d24fe186f9676430d2509d34523af7e32bf0c

rwaldron commented:
@soareschen can you do an "official review" please? Thanks!

soareschen commented:
Approved and merged. Hopefully this will fix https://twitter.com/taybrands/status/837423784256655360. :)
